File: .config/jp/tools/src/debug_app/trace_tests.rs
use super::{render, summarize};
use crate::util::trace::parse_lines;

/// One line exactly as the app writes it.
///
/// Pinned here and in `TraceTests.swift`, character for character.
/// Nothing else checks that the writer and the reader agree on the format: if
/// one of these two strings is edited alone, the other test is what says so.
const APP_LINE: &str = r#"{"timestamp":"2026-08-02T11:04:12.418293Z","level":"INFO","target":"JP.Transcript","fields":{"message":"transcript.render","duration_ms":84.219,"event_count":847,"footprint_mb":412},"spans":[{"name":"conversation.select"}]}"#;

#[test]
fn summarizes_a_line_the_app_wrote() {
    let summary = summarize(&parse_lines(APP_LINE));

    assert_eq!(summary.intervals, 1);
    assert_eq!(
        summary.slowest,
        Some(("transcript.render".to_owned(), 84.219))
    );
    assert_eq!(summary.footprint_mb, Some(412));
}

/// The block a snapshot prints beside its console delta.
#[test]
fn renders_the_slowest_interval_and_the_footprint_change() {
    let summary = summarize(&parse_lines(APP_LINE));

    assert_eq!(
        render(&summary, Some(374)).as_deref(),
        Some(
            "Trace, since the last call: 1 span. Slowest `transcript.render` 84 ms.\nFootprint \
             412 MB (+38 MB)."
        )
    );
}

/// The first snapshot of a run has nothing to compare against, and a `(+412
/// MB)` there would read as a spike rather than as the app's whole footprint.
#[test]
fn omits_the_change_when_nothing_was_reported_before() {
    let summary = summarize(&parse_lines(APP_LINE));

    assert_eq!(
        render(&summary, None).as_deref(),
        Some(
            "Trace, since the last call: 1 span. Slowest `transcript.render` 84 ms.\nFootprint \
             412 MB."
        )
    );
}

#[test]
fn counts_every_interval_and_keeps_the_longest() {
    let lines = [
        r#"{"timestamp":"2026-08-02T11:04:12.000000Z","level":"INFO","target":"JP.View","fields":{"message":"WorkspaceWindow.body","duration_ms":0.42,"footprint_mb":400}}"#,
        r#"{"timestamp":"2026-08-02T11:04:12.100000Z","level":"INFO","target":"JP.View","fields":{"message":"ConversationHistoryView.body","duration_ms":12.5,"footprint_mb":404}}"#,
        r#"{"timestamp":"2026-08-02T11:04:12.200000Z","level":"INFO","target":"JP.View","fields":{"message":"WorkspaceWindow.body","duration_ms":0.31,"footprint_mb":405}}"#,
    ]
    .join("\n");

    let summary = summarize(&parse_lines(&lines));

    assert_eq!(summary.intervals, 3);
    assert_eq!(
        render(&summary, Some(405)).as_deref(),
        Some(
            "Trace, since the last call: 3 spans. Slowest `ConversationHistoryView.body` 13 \
             ms.\nFootprint 405 MB."
        )
    );
}

/// The reference pair the app writes at startup times nothing, so a snapshot
/// that only saw it should not claim an interval.
#[test]
fn reports_an_origin_event_as_no_intervals() {
    let origin = r#"{"timestamp":"2026-08-02T11:04:10.000000Z","level":"INFO","target":"JP.Trace","fields":{"message":"trace.origin","mach_absolute_time":42000,"unix_time_ns":1785668650000000000,"timebase_numer":125,"timebase_denom":3}}"#;

    let summary = summarize(&parse_lines(origin));

    assert_eq!(summary.intervals, 0);
    assert_eq!(
        render(&summary, None).as_deref(),
        Some("Trace, since the last call: no intervals.")
    );
}

/// An app that traced nothing since the last call gets no block at all, rather
/// than a line saying so: the snapshot already carries three other sections.
#[test]
fn renders_nothing_when_the_delta_is_empty() {
    assert_eq!(render(&summarize(&parse_lines("")), Some(400)), None);
}

File: apps/macos/Tests/TraceTests.swift
import Foundation
import Testing

@testable import JP

/// One line exactly as the app writes it.
///
/// Pinned here and in `.config/jp/tools/src/debug_app/trace_tests.rs`, character
/// for character. Nothing else checks that the writer and the reader agree on
/// the format: if one of these two strings is edited alone, the other test is
/// what says so.
private let appLine = """
    {"timestamp":"2026-08-02T11:04:12.418293Z","level":"INFO","target":"JP.Transcript",\
    "fields":{"message":"transcript.render","duration_ms":84.219,"event_count":847,\
    "footprint_mb":412},"spans":[{"name":"conversation.select"}]}
    """

@Suite("Trace")
struct TraceTests {
    @Test("writes the line the tooling parses")
    func writesThePinnedLine() throws {
        let line = try #require(
            Trace.line(
                timestamp: "2026-08-02T11:04:12.418293Z",
                level: .info,
                target: "JP.Transcript",
                message: "transcript.render",
                fields: [
                    ("duration_ms", 84.219),
                    ("event_count", 847),
                    ("footprint_mb", 412),
                ],
                spans: ["conversation.select"]
            )
        )

        #expect(line == appLine)
    }

    /// The parser treats `spans` as optional, and most events are not nested
    /// inside anything.
    @Test("leaves the span stack out when there is none")
    func omitsAnEmptySpanStack() throws {
        let line = try #require(
            Trace.line(
                timestamp: "2026-08-02T11:04:10.000000Z",
                level: .info,
                target: "JP.Trace",
                message: "trace.origin",
                fields: [("timebase_numer", 125), ("timebase_denom", 3)],
                spans: []
            )
        )

        #expect(
            line == """
                {"timestamp":"2026-08-02T11:04:10.000000Z","level":"INFO","target":"JP.Trace",\
                "fields":{"message":"trace.origin","timebase_numer":125,"timebase_denom":3}}
                """
        )
    }

    /// RFC 3339, UTC, fractional seconds. A timestamp in local time or without
    /// the fraction still parses, and lands the event in the wrong place on a
    /// timeline drawn beside `jp`'s.
    @Test("formats timestamps as UTC to the microsecond")
    func formatsTimestamps() {
        #expect(
            Trace.timestamp(Date(timeIntervalSince1970: 1_785_668_652.418293))
                == "2026-08-02T11:04:12.418293Z"
        )
        #expect(
            Trace.timestamp(Date(timeIntervalSince1970: 0)) == "1970-01-01T00:00:00.000000Z")
    }

    @Test("reports the process footprint")
    func reportsTheFootprint() throws {
        let footprint = try #require(Trace.footprintMB())

        // A live process occupies something, and a footprint of hundreds of
        // gigabytes would mean the struct was read as the wrong shape.
        #expect(footprint > 0)
        #expect(footprint < 100_000)
    }

    @Test("converts mach ticks to milliseconds")
    func convertsMachTicks() {
        let timebase = MachTimebase.current
        // Exactly one second's worth of ticks, whatever this machine counts in.
        let ticks =
            UInt64(1_000_000_000) * UInt64(timebase.denominator)
            / UInt64(timebase.numerator)

        #expect(abs(Trace.milliseconds(ticks) - 1000) < 0.01)
    }
}

/// Nested in ``WorkspaceSuite`` because these read `JP_DEBUG_STATE_DIR`, which
/// the whole process shares.
extension WorkspaceSuite {
    @Suite("TraceWriter")
    struct TraceWriterTests {
        /// The shipping configuration, and the one thing this must never get
        /// wrong: an installed app writing a trace file into someone's disk
        /// would be a defect, not a feature.
        @Test("creates nothing without a state directory")
        func createsNothingWithoutADirectory() throws {
            let root = try makeTemporaryDirectory()
            defer { removeSandbox(root) }

            #expect(TraceWriter(directory: nil, fileName: Trace.fileName) == nil)
            #expect(try FileManager.default.contentsOfDirectory(atPath: root.path).isEmpty)
        }

        /// The process-wide sink, resolved from the environment the test host
        /// runs under. Serialized with the tests that set that variable, so it
        /// is unset here.
        @Test("records nothing when the app was launched as it ships")
        func recordsNothingWhenLaunchedNormally() {
            #expect(ProcessInfo.processInfo.environment[DebugState.variable] == nil)
            #expect(Trace.isRecording == false)
            #expect(Trace.url == nil)

            // Reaches every sink there is. Nothing to assert but that it neither
            // crashes nor has a file to write to.
            Trace.event("test.event")
            Trace.interval("test.interval").end()
        }

        @Test("appends one line per event")
        func appendsOneLinePerEvent() throws {
            let root = try makeTemporaryDirectory()
            defer { removeSandbox(root) }

            let writer = try #require(TraceWriter(directory: root, fileName: Trace.fileName))
            writer.append("first")
            writer.append("second")

            #expect(writer.url == root.appendingPathComponent("trace.jsonl"))
            #expect(try String(contentsOf: writer.url, encoding: .utf8) == "first\nsecond\n")
        }

        /// The directory a harness names does not exist yet when it launches the
        /// app into it.
        @Test("creates the directory it was pointed at")
        func createsTheDirectory() throws {
            let root = try makeTemporaryDirectory()
            defer { removeSandbox(root) }
            let state = root.appendingPathComponent("state")

            let writer = try #require(TraceWriter(directory: state, fileName: Trace.fileName))
            writer.append("line")

            #expect(try String(contentsOf: writer.url, encoding: .utf8) == "line\n")
        }

        /// A relaunch truncates the file, and a second writer on the same path
        /// must not overwrite what the first one wrote.
        @Test("appends to a file that already has lines in it")
        func appendsToAnExistingFile() throws {
            let root = try makeTemporaryDirectory()
            defer { removeSandbox(root) }

            let first = try #require(TraceWriter(directory: root, fileName: Trace.fileName))
            first.append("first")

            let second = try #require(TraceWriter(directory: root, fileName: Trace.fileName))
            second.append("second")

            #expect(try String(contentsOf: second.url, encoding: .utf8) == "first\nsecond\n")
        }
    }
}
